js/bundle.go

```diff
@@ -307,6 +310,48 @@ func (b *Bundle) instantiate(logger logrus.FieldLogger, rt *goja.Runtime, init *
 	}
 	rt.Set("__ENV", env)
 	rt.Set("__VU", vuID)
+	rt.SetPromiseRejectionTracker(func(p *goja.Promise, op goja.PromiseRejectionOperation) {
+		// TODO this only handles the case where a promise get's rejected (either by an exception or by rejecting it
+		// programmatically). But the other case where for an already rejected promise (one that we've logged the error)
+		// there is now a reject handler attached is not handled.
+		// Unfortunately that might mean that we log the warning to the user at much later time (you can set a reject on
+		// a promise even a few loops of the event loop later.
+		// Firefox seems to manage to catch ones that happen before the end of the currently executing script and
+		// doesn't log the once that get a rejection handler attached. We can do something similar although given that a
+		// script can go for much longer than most script executions on firefox (and also this is just my observation)
+		// we should probably have it time limited as otherwise we will be keeping references to rejected promises that
+		// might never get a rejection handler and also it will be better to not log about this 5seconds after it has
+		// happened.
+		// Read Notes on https://tc39.es/ecma262/#sec-host-promise-rejection-tracker
+		if op == goja.PromiseRejectionReject {
+			frames := rt.CaptureCallStack(20, nil)
+			buf := &bytes.Buffer{}
+			for _, frame := range frames {
+				frame.Write(buf)
+				buf.WriteRune('\n')
+			}
+
+			logger.Warnf("Uncaught (in Promise) %s\n %s", p.Result(), buf.String())
+		}
+	})
+	_ = rt.Set("setTimeout", func(f func() error, t float64) {
+		// TODO checks and fixes
+		// TODO maybe really return something to use with `clearTimeout
+		// TODO support arguments ... maybe
+		runOnLoop := init.loop.reserve()
```

Comment:
I think I get it now: loop.reserve() essentially returns a "runner" function that will execute whatever closure you pass to it whenever the eventloop has a slot available. Did I get this correctly?

In this code specifically, you say: reserve a spot on the event loop, give me a handler that I can use to execute some closure. Then start a goroutine with a timer set to whatever was passed to `setTimeout` in the JS script, and when the time ticks, use the handler to execute the closure/callback passed to `setTimeout`.

Sorry if that's noise, but it helps me verify I really understand what's happening :)

Reply:
Yes, I like to think of it  as a delayed enqueue/add on the event loop that makes it, so the event loop doesn't exit before it happens.